odl/tomo/geometry/conebeam.py

```diff
@@ -1236,6 +1238,178 @@ def frommatrix(cls, apart, dpart, src_radius, det_radius, init_matrix):
+    between the linear and angular sampling intervals are
+
+    .. math::
+        \\Delta s = R \\Delta \\alpha, \quad R^2 = r^2 + (w / 2)^2,
+
+    where :math:`w` is the width of the detector.
+    Thus, to satisfy the angular detector condition one can choose
+
+    .. math::
+        \\Delta s \leq \\frac{\\pi \sqrt{r^2 + (w / 2)^2}}{r \\Omega}.
+
+    The geometry returned by this function satisfies these conditions exactly.
+
+    If the domain is 3-dimensional, a circular cone beam geometry is
+    created with the third coordinate axis as rotation axis. This does,
+    of course, not yield complete data, but is equivalent to the
+    2D fan beam case in the :math:`z = 0` slice.
+
+    References
+    ----------
+    .. _Mathematical Methods in Image Reconstruction: \
+http://dx.doi.org/10.1137/1.9780898718324
+    """
+    # Find maximum distance from rotation axis
+    corners = space.domain.corners()[:, :2]
+    rho = np.max(np.linalg.norm(corners, axis=1))
+
+    # Find default values according to Nyquist criterion.
+
+    # We assume that the function is bandlimited by a wave along the x or y
+    # axis. The highest frequency we can measure is then a standing wave with
+    # period of twice the inter-node distance.
+    min_side = min(space.partition.cell_sides[:2])
+    omega = np.pi / min_side
+
+    # Compute minimum width of the detector to cover the object. The relation
+    # used here is (w/2)/(rs+rd) = rho/rs since both are equal to tan(alpha),
+    # where alpha is the half fan angle.
+    rs = float(src_radius)
+    rd = float(det_radius)
+    r = src_radius + det_radius
+    w = 2 * rho * (rs + rd) / rs
+
+    # Compute minimum number of pixels given the constraint on the
+    # sampling interval and the computed width
+    rb = np.hypot(r, w / 2)  # length of the boundary ray to the flat detector
+    num_px_horiz = int(np.ceil(w * omega * r / (2 * np.pi * rb)))
+    delta_s = w / num_px_horiz
+
+    if space.ndim == 2:
+        det_min_pt = -w / 2
+        det_max_pt = w / 2
+        if det_shape is None:
+            det_shape = num_px_horiz
+    elif space.ndim == 3:
+        # Compute number of vertical pixels required to cover the object,
+        # using the same sampling interval vertically as horizontally.
+        # The reasoning is the same as for the computation of w.
+        corners = space.domain.corners()
+        rho_3d = np.max(np.linalg.norm(corners, axis=1))
+        h = 2 * rho_3d * (rs + rd) / rs
+        num_px_vert = int(np.ceil(h / delta_s))
+        h = num_px_vert * delta_s  # make multiple of delta_s
+
+        det_min_pt = [-w / 2, -h / 2]
+        det_max_pt = [w / 2, h / 2]
+        if det_shape is None:
+            det_shape = [num_px_horiz, num_px_vert]
+
+    if num_angles is None:
+        num_angles = int(np.ceil(2 * omega * rho) * r / (r + rho))
+
+    angle_partition = nonuniform_partition(
+        np.linspace(0, np.pi, num_angles, endpoint=False))
```

Comment:
In a cone-beam geometry this will not work. The angles have to go from 0 to pi+fan_angle for full data, however in that case `parker_weighting` is needed. To work without weighting you need to go from 0 to 2pi.